feat: create base wflow model

LuukBlom opened the pull request:
## Issue Addressed  
Fixes #346  

## What Has Changed  
This PR introduces a new base class, `WflowModelBase`, that consolidates the shared functionality between `WflowModel` and `WflowSedimentModel`. (class names are subject to change)

`WflowModelBase` contains the shared **components**, **properties**, and **generic setup/model functions**, while `WflowModel` and `WflowSedimentModel` only need to implement their model-specific setup methods and upgrade logic.  

### WflowModelBase  
Includes:  
- **Components**: `config`, `forcing`, `geoms`, `states`, `staticmaps`, `tables`  
- **Properties**: `basins`, `basins_highres`, `rivers`  
- **Setup functions**: `setup_config`, `setup_basemaps`, `setup_lulcmaps`, `setup_lulcmaps_from_vector`, `setup_lulcmaps_with_paddy`, `setup_laimaps`, `setup_laimaps_from_lulc_mapping`, `setup_outlets`, `setup_config_output_timeseries`, `setup_gauges`, `setup_areamap`, `setup_riverwidth`, `setup_constant_pars`, `setup_grid_from_raster`
- **Model functions**: `write`, `read`, `read_<component>`, `write_<component>`, `set_root`, `get_config`, `set_config`, `clip`, `flwdir`  

### WflowModel (inherits from `WflowModelBase`)  
Adds SBM-specific setup functions:  
`setup_rivers`, `setup_floodplains`, `setup_reservoirs_no_control`, `setup_reservoirs_simple_control`, `setup_soilmaps`, `setup_ksathorfrac`, `setup_ksatver_vegetation`, `setup_lulcmaps_with_paddy`, `setup_glaciers`, `setup_precip_forcing`, `setup_temp_pet_forcing`, `setup_pet_forcing`, `setup_rootzoneclim`, `setup_1dmodel_connection`, `setup_allocation_areas`, `setup_allocation_surfacewaterfrac`, `setup_domestic_demand`, `setup_irrigation`, `setup_irrigation_from_vector`, `setup_cold_states`  

Also implements `upgrade_to_v1_wflow`.  

### WflowSedimentModel (inherits from `WflowModelBase`)  
Adds sediment-specific setup functions such as:  
`setup_rivers`, `setup_natural_reservoirs`, `setup_reservoirs`, `setup_outlets`, `setup_gauges`, `setup_lulcmaps`, `setup_lulcmaps_from_vector`, `setup_riverbedsed`, `setup_canopymaps`, `setup_soilmaps`  

Also implements `upgrade_to_v1_wflow`.  

## Checklist  
- [x] Updated tests or added new ones  
- [x] Branch is up to date with `main`  
- [x] Tests & pre-commit hooks pass  
- [x] Documentation updated if needed

savente93 commented on hydromt_wflow/wflow.py:
I think that is a good idea

savente93 reviewed (approved):
I'm assuming there's not really new code, it's just moved around. In that case I think this looks good. I'm wondering if `setup_pet_forcing` doesn't belong in the base, but I'll leave that up to Hélène. Other than that, I think this looks good. 

Personally I think renamingto WflowSbmModel is better, but I'll also leave that up to our domain expert ;)

hboisgon reviewed (commented):
Shame the history of the wflow_sbm file got completely lost. This makes it difficult to review... So here is a copy-paste and edits of your very nice PR message (thanks for that!) of what should be where and what is still todo. I will blindly trust that you will implement this list (better if you could keep the history of wflow_sbm but guess that will be difficult?)

For the rest I really like this PR, nice job :) a few things:

- [ ] maybe the Base class could be named ``WflowBaseModel`` rather than ``WflowModelBase``? But I don't mind too much.
- [ ] ``WflowModel`` or ``WflowSbmModel``. Let me check with the Wflow developers. We have a wflow_gwf concept and there was plan to still have wflow_flextopo. If that is the case I would lean towards ``WflowSbmModel``. This means a lot of changes but users will already have to change all their calls to config, grid to conffig.data and staticmaps.data so this one change may not be the end of the world.
- [ ] clip_rivers functionality should be split. I think for the manning coefficient you can actually move it to its own separate setup_method. It's been on a wishlist for a while and this means setup_rivers would be exactly the same for sbm and sediment.
- [ ] clip_staticmaps: the reservoir part should be moved out as it is sbm specific. Technically the same should be done for sediment but with different parameter name. We can also decide to add the parameter list as a function argument in Base and then sbm and sediment can provide their own list.

Okay now which function should be where:

### WflowModelBase  
Includes:  
- **Components**: `config`, `forcing`, `geoms`, `states`, `staticmaps`, `tables`  
- **Properties**: `basins`, `basins_highres`, `rivers` , `flwdir` 
- **Setup functions** (in the right order!): `setup_config`, `setup_config_output_timeseries`, `setup_basemaps`, `setup_rivers`, `setup_riverwidth`, `setup_lulcmaps`, `setup_lulcmaps_from_vector`,  `setup_outlets`,  `setup_gauges`, `setup_constant_pars`, `setup_grid_from_raster`,`setup_areamap`
- **Model functions**: `write`, `read`, `read_<component>`, `write_<component>`, `set_root`, `get_config`, `set_config`, `clip` 

### WflowModel (inherits from `WflowModelBase`)  
Adds SBM-specific setup functions (if the order/history could be kept that would be great! here is a random order):  
`setup_river_roughness` (new: extract manning from setup_rivers), `setup_floodplains`, `setup_reservoirs_no_control`, `setup_reservoirs_simple_control`,`setup_lulcmaps_with_paddy`, `setup_laimaps`, `setup_laimaps_from_lulc_mapping`, `setup_soilmaps`, `setup_ksathorfrac`, `setup_ksatver_vegetation`, `setup_glaciers`, `setup_precip_forcing`, `setup_precip_from_point_timeseries`, `setup_temp_pet_forcing`, `setup_pet_forcing`, `setup_rootzoneclim`, `setup_1dmodel_connection`, `setup_allocation_areas`, `setup_allocation_surfacewaterfrac`, `setup_domestic_demand`, `setup_irrigation`, `setup_irrigation_from_vector`, `setup_cold_states` , `upgrade_to_v1_wflow`.  

### WflowSedimentModel (inherits from `WflowModelBase`)  
Adds sediment-specific setup functions such as:  
`setup_natural_reservoirs`, `setup_reservoirs`, `setup_outlets`, `setup_gauges`, `setup_lulcmaps`, `setup_lulcmaps_from_vector`, `setup_riverbedsed`, `setup_canopymaps`, `setup_soilmaps` , `upgrade_to_v1_wflow`.

LuukBlom commented on hydromt_wflow/wflow.py:
Yeah, sorry I dont know what happened to the history. I will implement your comments and then try and see if I can resolve the history! If not, I'll make sure its exactly like your comments are saying. (Also, i didnt delete any code in this PR, its just renames and moving stuff around, so nothing gets deleted!)

LuukBlom replied:
Ok I just spent quite some time trying to rewrite the history of this branch to try to keep the old file/line history, but could not get it to work and am ready to give up hahah. 
Git is apparently known for being bad splitting a large file into smaller ones. 

I tried all of the below: 
https://dev.to/deckstar/how-to-git-copy-copying-files-while-keeping-git-history-1c9j
https://stackoverflow.com/questions/38729453/how-to-preserve-git-history-when-refactoring-into-multiple-files
https://esmithy.net/2020/08/15/preserve-git-blame-history-refactoring/

LuukBlom commented on docs/getting_started/example_index.rst:
```diff
@@ -36,3 +36,28 @@ For a static (non-interactive) view of the examples follow one of the links belo
 *   `Upgrade to Wflow.jl version 1 <../_examples/upgrade_to_wflow_v1.ipynb>`_
+
+
+Examples (static view)
+----------------------
```
I was running the docs-dummy task to check if I broke the docs, and it was complaining the examples were not part of a TOC (somehow?). 
Adding this resolved it, but I can delete as well. I will have a look again after implementing your comments to see if it is actually still required.

LuukBlom commented on hydromt_wflow/wflow.py:
```diff
@@ -82,327 +69,20 @@ def __init__(
-            self.set_config("model.cell_length_in_meter__flag", True)
-
     @hydromt_step
     def setup_rivers(
```
basemodel now has setup_rivers without the manning part. 
SbmModel now has setup_river_roughness, which calls super().setup_rivers and then the removed manning part from the base.

LuukBlom commented:
Current state of the PR:

All functions are in the correct place according to https://github.com/Deltares/hydromt_wflow/pull/535#pullrequestreview-3167033737
I could not salvage the history unfortunately, see https://github.com/Deltares/hydromt_wflow/pull/535#discussion_r2309943561. 

Non-trivial changes: (e.g. more than move/delete/copy): 
- added `write_region` to `WflowBaseModel.write()`. This was needed to actually write the region file (this was no longer being done due to our `region_filename` changes in the `__init__`) so it can be read and compared in test_model_build. 

- `clip_grid`: changes according to https://github.com/Deltares/hydromt_wflow/pull/535#discussion_r2308923668. `clip_grid` is now split into `WflowBaseModel` (excluding reservoirs part) and `WflowSbmModel` (calls `super().clip_grid()` and then does the reservoirs part). TODO what to do with sediment?

- `clip_states`: `WflowBaseModel` has the generic clipping. `WflowSbmModel` overwrites `clip_states`, adding the non-generic part (remove_maps). TODO what to do with sediment?

- `setup_rivers`: moved to `WflowBaseModel`, excluding the manning part. `WflowSbmModel` now has `setup_river_roughness`, which calls `super().setup_rivers` and then adds the manning roughness.

hboisgon reviewed (changes requested):
Thanks Luuk for the changes! Shame the history could not be kept and sorry you spent a lot of time on it, was not what I meant... (I thought you might have moved things around instead of just deleting)

I think most of the PR is there, thanks for all the changes you made! The last thing that still needs modification is the setup_rivers method and the clipping. Let me know if anything is unclear for setup_rivers. There is a lot of duplications in the current implementation that we could avoid if we restructure the function a little.

LuukBlom commented on docs/api.rst:
```diff
@@ -255,8 +253,8 @@ Other general methods
 
 .. _data_containers:
 
-WflowModel components
```
Sounds good! Should I also remove the duplicate setup_functions into a section of WflowBaseModel?

LuukBlom commented on docs/changelog.rst:
Agreed, this happened accidentally during global search and replace so I will revert!

LuukBlom commented on hydromt_wflow/wflow_base.py:
```diff
@@ -0,0 +1,2490 @@
+            >> self.set_config('b', 'c', 'd', 99) # identical to set_config('b.d.e', 99)
+            >> {'a': 1, 'b': {'c': {'d': 99}}}
+        """
+        key = ".".join(args[:-1])
+        value = args[-1]
+        self.config.set(key, value)
```
What do you mean with something different?

reading this again now, I think we should add a check that len(args) is at least 2. (as mentioned in the docstring as well, but the implementation does not check it so `WflowBaseModel.set_config("test")` actually calls `self.config.set("test", "test")` without an error.

LuukBlom commented on docs/getting_started/faq.rst:
```diff
@@ -17,7 +17,7 @@ Building a Wflow model
 This ensures that all upstream contributing cells are well taken into account. Using other options might cause your Wflow model
 run crash. If you are sure of what you are doing, you can optionally use the ``geom`` option and provide an exact shapefile of the
-basin/subbasins you want to derive, that was prepared using the same DEM source that you intend to use in the :py:func:`~WflowModel.setup_basemaps`
+basin/subbasins you want to derive, that was prepared using the same DEM source that you intend to use in the :py:func:`~WflowBaseModel.setup_basemaps`
```
For now I have added WflowBaseModel to the api docs.

I also added a line to the docstring of the class where we mention you shouldnt use this class directly and you should only use its subclasses.
It is also not a hydromt-entrypoint so it is not accessible via cli.

Nevertheless, I added this to the `WflowBaseModel.__init__`:
```python
class WflowBaseModel:
  def __init__(self, ...):
    ...
    if type(self) is WflowBaseModel:
      raise ValueError(
        "``WflowBaseModel`` is an abstract class and cannot be instantiated "
        "directly. Please use one of its subclasses defined as hydromt-entry "
        "points: [``WflowSbmModel``, ``WflowSedimentModel``]"
      )
    ...
```
This solution is very simple and prevents any incorrect usage, while still allowing the subclasses to inherit from `WflowBaseModel`.
Alternatively, we could make `WflowBaseModel` inherit from `ABC`, but that means we introduce multiple inheritance, which I dont like tbh.

LuukBlom commented:
Current state of the PR:

**History**
I could not salvage the history unfortunately, see [this comment](https://github.com/Deltares/hydromt_wflow/pull/535#discussion_r2309943561) for what I tried.

**Function locations & order**
All functions are in the correct location and order according to https://github.com/Deltares/hydromt_wflow/pull/535#pullrequestreview-3167033737. 
Except:
- `setup_other_demand` was missing from the comment, so I put the function after `setup_domestic_demand_from_population` and before `setup_irrigation` in `WflowSbmModel`
, see below also.


**Non-trivial changes: (e.g. more than move/delete/copy/docstring/formatting):** 
- added `write_region` to `WflowBaseModel.write()`. This was needed to actually write the region file (this was no longer being done due to us setting all `region_filename` to `None` in the `__init__`). Added it so the tests (`test_model_build`) are green, but we can remove.

- `clip_grid`: changes according to https://github.com/Deltares/hydromt_wflow/pull/535#discussion_r2308923668. `clip_grid` is now split into `WflowBaseModel` (excluding reservoirs part) and `WflowSbmModel` (calls `super().clip_grid()` and then does the reservoirs part).

- `clip_states`: `WflowBaseModel` has the generic clipping. `WflowSbmModel` overwrites `clip_states`, adding the non-generic part (remove_maps).

- `setup_rivers`: changes according to https://github.com/Deltares/hydromt_wflow/pull/535#discussion_r2308895806.  `WflowBaseModel.setup_rivers`: extracted manning part, created `WflowBaseModel.setup_rivers`, `WflowSbmModel.setup_rivers` and `WflowSbmModel.setup_river_roughness`

- `WflowSbmModel` now has `setup_river_roughness`, which requires `setup_basemaps` and `setup_rivers` to be called before. 

- updated cli name for sbm model from `wflow` to `wflow_sbm`

LuukBlom commented on tests/components/test_config_component.py:
```diff
@@ -190,7 +190,7 @@ def test_wflow_config_component_read_default_write_mode(
     component = WflowConfigComponent(
         model=mock_model,
-        default_template_filename=Path(DATADIR, "wflow", "wflow_sbm.toml"),
+        default_template_filename=str(DATADIR / "wflow_sbm" / "wflow_sbm.toml"),
```
I put in `WflowBaseModel.__init__` some code that generates this path using the `name` attribute of the Model. Which allows us to only define the name for subclasses, and then `super().__init__()` takes care of constructing the path and passing it to the `ConfigComponent`.

```python
class WflowBaseModel(...):
    name: str = "wflow"

    def __init__(...):
        ...
        default_filename = self._DATADIR / self.name / f"{self.name}.toml"
        ...

class WflowSbmModel(...):
    name: str = "wflow_sbm"
    ...
```
